moderntribe/moose
# Feature: 1Password CLI Integration

## What does this do/fix?
This work swaps our our manual process of creating composer auth.json files via copy-paste and the json object blob in the `composer_auth_json` GitHub secret for an automated process via the 1Password CLI. This automation works for both local dev and GitHub actions.

For local dev, I've created a composer script generates the auth.json file from the template using the local developer's 1Pass account. (This requires the [1Password CLI](https://developer.1password.com/docs/cli/get-started) to be installed and configured for a user first.)

For GH Actions, a specific [1Pass Service Account](https://developer.1password.com/docs/service-accounts/) is used to authenticate against the required vault.

This PR also updates our documentation in GitHub to be more robust and concise. 

To do:
- [ ] Create a demo video of this feature
- [x] Define & test env-specific secrets for GH Actions
  Confirmed that setting an environmental secret to override the repository secrets works. It requires that the workflow jobs utilizing the env-specific secrets define an `environment` that matches the Environment named in the repository configuration.
- [x] Update Readme docs with the new project setup process
- [x] Write up complete documentation on the 1Password integration 
- [x] ~Update deploy actions with the new composer install action~
- [x] How best to handle 1Password service accounts for GHA integration.
  - 1Password Service Accounts can't be modified and can't add new vault access automatically, so a new Service account for each vault that needs to be accessed should be created.
- [x] How best to handle different license key values from clients.
  1. Create a new 1Pass item in the project's vault using the MT-Composer-Auth item as a template.
  2. Create a new 1Password Service Account with read access to the project's vault.
  3. Create project-specific GitHub secrets to hold the correct 1Pass Service Account Token (`OP_SERVICE_ACCOUNT_TOKEN`), vault (`OP_VAULT`), and item (`OP_ITEM`).
  4. Future Improvement: One thing I haven't examined closely is how we might use different secrets per environment with the same projects such that deploying to Dokku can use the MT vault items and the client's env's can use their own vault items.
- [x] How best to handle additions or changes to the `auth.template.json file`. For example, how to add `github-oauth` token and other new auth values (Yoast SEO, etc) with the least amount of friction. 
  1. Add a new section and item(s) to the specific project's 1pass item
  2. Update the `auth.template.json` file with the path to the new section & item(s)
- [x] ~Ideally: How might we automate the script run for local dev. This needs to run on the local machine, outside of the Lando containers, as it requires the user's 1Pass account be setup to auth against the vault.~
  - There is no good way to do this due to how Lando is designed to run all actions within the containers it spawns. I've modified the project setup instructions such that the first command to run is `composer create-auth`, before running `lando start`.
